pymc4/distributions/state_functions.py

```diff
@@ -0,0 +1,250 @@
+        """
+        Proposal function that is passed as the argument
+        to RWM kernel
+
+        Parameters
+        ----------
+        state_parts : List[tf.Tensor]
+            A list of `Tensor`s of any shape and real dtype representing
+            the state of the `current_state` of the Markov chain
+        seed: Optional[int]
+            The random seed for this `Op`. If `None`, no seed is
+        applied
+            Default value: `None`
+
+        Returns
+        -------
+        List[tf.Tensor]
+            A Python `list` of The `Tensor`s. Has the same
+            shape and type as the `state_parts`.
+
+        Raises
+        ------
+        ValueError: if `scale` does not broadcast with `state_parts`.
+        """
+        pass
+
+    @abc.abstractmethod
+    def __eq__(self, other) -> bool:
+        """
+        Comparison operator overload of each proposal sub-class.
+        The operator is required to disnguish same proposal functions to separate
+        samplers in `Compound step`
+
+        Parameters
+        ----------
+        other: pm.distributions.Proposal
+        Another instance of `Proposal` sub-class.
+
+        Returns
+        -------
+        bool
+            True/False for equality of instances
+        """
+        pass
+
+    def __call__(self):
+        return self._fn
+
+
+class CategoricalUniformFn(Proposal):
+    """
+    Categorical proposal sub-class with the `_fn` that is sampling new proposal
+    from catecorical distribution with uniform probabilities.
+
+    Parameters
+    ----------
+    classes: int
+        Number of classes for catecorical distribution
+    name: Optional[str]
+        Python `str` name prefixed to Ops created by this function.
+        Default value: 'categorical_uniform_fn'.
+    """
+
+    _name = "categorical_uniform_fn"
+
+    def __init__(self, classes: int, name: Optional[str] = None):
+        super().__init__(name)
+        self.classes = classes
+
+    def _fn(self, state_parts: List[tf.Tensor], seed: Optional[int]) -> List[tf.Tensor]:
+        with tf.name_scope(self._name or "categorical_uniform_fn"):
+            deltas = tf.unstack(
+                tf.map_fn(
+                    lambda x: tfd.Categorical(logits=tf.ones(self.classes)).sample(
+                        seed=seed, sample_shape=tf.shape(x)
+                    ),
+                    tf.stack(state_parts),
+                )
+            )
+            return deltas
+
+    def __eq__(self, other) -> bool:
+        return self._name == other._name and self.classes == other.classes
+
+
+class BernoulliFn(Proposal):
+    """
+    Bernoulli proposal sub-class with the `_fn` that is sampling new proposal
+    from bernoulli distribution with p=0.5.
+
+    Parameters
+    ----------
+    scale: Union[List[Any], Any]
+        a `Tensor` or Python `list` of `Tensor`s of any shapes and `dtypes`
+        controlling the scale of the proposal distribution.
```

Comment:
Why is scale needed in Bernoulli random?

Reply:
Yeap, it is not needed. Sorry, in previous design I had `scale` as the requirement for all proposal functions.